accepted/0000-caasp-integration.md

```diff
@@ -0,0 +1,395 @@
+
+### Deploy a cluster using Uyuni/SUSE Manager
+
+Under Uyuni/SUSE Manager main menu, a new menu item "Clusters" should be introduced.
+All deployable and manageable cluster products (e.g. CaaSP, SOC, SES) will be in the sub-menu.
+
+Under that item, it should be possible to create, edit, and remove a cluster.
+
+To create a cluster, a user has to enter its name.
+
+Every detail about the cluster is saved into corresponding database tables (e.g. `susecluster`).
+
+API support for all the activities described below should be implemented as well as front-end support.
+
+#### Cluster requirements
+
+If the cluster required a specific set of nodes optionally with a defined configuration, the cluster provider must offer a way to deploy the whole infrastructure automatically, e.g. using AutoYaST profiles.
+
+Uyuni/SUSE Manager will provision bare-metal machines using the supplied AutoYaST profile to have a complete set of machines ready to be hosting the cluster.
+
+#### The Management Node
+
+The cluster provider must also indicate:
+
+- how and the number of Management Node(s) must be elected among the provisioned machines (e.g. based on hardware requirements)
+- how Management Node(s) should be deployed (e.g. additional software to be installed via AutoYaST or Salt states)
+- whether Management Node(s) can be managed by Uyuni/SUSE Manager (specifically: can users manage the updates of the Management Node via Uyuni/SUSE Manager?)
+
+#### Cluster bootstrapping
+
+The cluster must be initiated by the Management Node that has all the tools to bootstrap and manage the cluster. Specific instructions on how to bootstrap the cluster must be provided.
+Questions that need to be answered by the cluster provider:
+- Upon successful bootstrapping, can the worker nodes be managed by Uyuni/SUSE Manager?
+- Do the nodes have a specific agent running on them and/or can we install a `salt-minion` or use `salt-ssh` on those nodes?
+- Can the nodes be managed by Uyuni/SUSE Manager? Can the packages be updated or do they require specific handling by the Management Node?
+
+### Manage a cluster using Uyuni/SUSE Manager
+
+Uyuni/SUSE Manager offers to show the status of the running cluster by either accessing the cluster directly or via the Management Node.
+In this regard, the cluster provider should provide a way to expose the cluster status (e.g. running nodes, the health of each node, monitoring metrics...) to Uyuni/SUSE Manager.
+
+#### Add and removal of nodes to the cluster
+
+Uyuni/SUSE Manager also offers the possibility of adding and removing nodes to the cluster:
+
+- To add a node from scratch, an AutoYaST profile can be reused
+- To remove a node, Uyuni/SUSE Manager must remove the node from the cluster using the Management Node and issuing the proper procedure
+
+Uyuni/SUSE Manager should be instructed to perform additional actions depending on the role of added/removed node (e.g. control plane, worker, load balancer, ...).
```

Comment:
> Does that mean that SUSE Manager now holds all the information on how to deploy and manage things? What about the customers who are not using SUSE Manager and want to deploy a product relying on caasp on bare metal, like SES?
> 
> Assuming SUMA is the unique holder of that information, how will the SUMA team deal with the amount of work that will fall onto them in a timely fashion? Will this be delegated to the appropriate teams? What will the interface be?

Customers will be able to use standardized deployments even when not using SUSE Manager.

Reply:
This doesn't answer my question.

It seems the phrase is written around the fact SUSE manager will have to know how do some some actions based on role of nodes. It means you will need to gather knowledge from the different teams, and merge it into the product.

Does that make sense? Do you really want to have domain knowledge of all products? If yes, how will you make sure 1) the code of different products don't overlap, 2) that you will provide enough bandwidth to merge domain specific work from different teams in a timely fashion, ... ?

What I mean here, is that it's probably best to define an interface for which other teams can based themselves on.
I suppose for CaaSP/SES/SOC, that interface could that teams for CaaSP/SES/SOC need to deploy their tooling using k8s native concepts (because those can be deployed by any lifecycle tooling of the k8s world, without any impact on SUMA).

Long story short, that phrase isn't clear about what you will implement or request from others, at least to me.

Reply:
I am starting from the example of CaaS Platform:
- Deployment:
  - CaaS Platform provided a set of AutoYaST profiles (that can be customizable) to deploy cluster nodes from bare metal
  - CaaS Platform provided `skuba`, a tool to bootstrap those nodes into the final cluster
- Management:
  - CaaS Platform provided `skuba` and `skuba-update`: these tools cover the standard operations for managing the cluster

In this regard, Uyuni/SUSE Manager must know how to:
- deploy bare metal using AutoYaST (already capable)
- interact with `skuba` and `skuba-update` (can be done via Salt)

The final outcome is that every interfacing cluster provider must provide these two pieces of information: a deployment method (possibly already supported in SUSE Manager) and a way to bootstrap and manage the cluster (in a tool provided by the cluster provider).
Does this clarify the statement? Should I insert into the RFC text?

Reply:
> It seems the phrase is written around the fact SUSE manager will have to know how do some some actions based on role of nodes. It means you will need to gather knowledge from the different teams, and merge it into the product.

In a sense, yes.

Each individual product must provide an independent installer, and should provide some kind of API, command-line tool or way for SUSE Manager to ask it to install. SUSE Manager should only have the minimum knowledge needed to be able to install, show status, audit, gather metrics, update, etc other SUSE products. That's the reason for the "cluster management" concept: we are not going to reimplement CaaSP features in SUSE Manager but we want to drive CaaSP from SUSE Manager.

> What I mean here, is that it's probably best to define an interface for which other teams can based themselves on.
> I suppose for CaaSP/SES/SOC, that interface could that teams for CaaSP/SES/SOC need to deploy their tooling using k8s native concepts (because those can be deployed by any lifecycle tooling of the k8s world, without any impact on SUMA).

Why Kubernetes? As I said in another comment, there's more clusters than Kubernetes.

Reply:
Yes, we agree with you: SUMA does need to know what is running on top of K8S.